service/src/main/java/bio/terra/workspace/service/workspace/flight/CreateCustomGcpRolesStep.java

```diff
@@ -31,37 +37,54 @@ public CreateCustomGcpRolesStep(IamCow iamCow) {
-        logger.info(
-            "Creating role {} with permissions {} in project {}",
-            customRole.getRoleName(),
-            customRole.getIncludedPermissions(),
-            projectId);
-        iamCow.projects().roles().create("projects/" + projectId, request).execute();
-      } catch (GoogleJsonResponseException googleEx) {
-        // Because this step may run multiple times, we need to handle duplicate role creation.
-        // The project was retrieved from RBS earlier in this flight, so we assume that any conflict
-        // of role names must be due to duplicate step execution.
-        if (googleEx.getStatusCode() == HttpStatus.CONFLICT.value()) {
-          continue;
-        } else {
-          throw new RetryException(googleEx);
-        }
-      } catch (IOException e) {
-        // Retry on IO exceptions thrown by CRL.
-        throw new RetryException(e);
-      }
+    // First, create the project-level custom roles.
+    // Multiple WSM roles may share the same GCP role. De-duping here prevents duplicate requests,
+    // which would lead to unnecessary CONFLICT responses from GCP.
+    ImmutableSet<CustomGcpIamRole> customProjectRoles =
+        CloudSyncRoleMapping.CUSTOM_GCP_PROJECT_IAM_ROLES.values().stream()
+            .collect(ImmutableSet.toImmutableSet());
+    for (CustomGcpIamRole customProjectRole : customProjectRoles) {
+      createCustomRole(customProjectRole, projectId);
+    }
+    // Second, create the resource-level custom roles.
+    for (CustomGcpIamRole customResourceRole :
+        CustomGcpIamRoleMapping.CUSTOM_GCP_RESOURCE_IAM_ROLES.values()) {
+      createCustomRole(customResourceRole, projectId);
     }
     return StepResult.getStepResultSuccess();
   }
 
+  /**
+   * Utility for creating custom roles in GCP from  WSM's CustomGcpIamRole objects.
+   */
+  private void createCustomRole(CustomGcpIamRole customRole, String projectId)
+      throws RetryException {
+    try {
+      Role gcpRole =
+          new Role()
+              .setIncludedPermissions(customRole.getIncludedPermissions())
+              .setTitle(customRole.getRoleName());
+      CreateRoleRequest request =
+          new CreateRoleRequest().setRole(gcpRole).setRoleId(customRole.getRoleName());
+      logger.info(
+          "Creating role {} with permissions {} in project {}",
+          customRole.getRoleName(),
+          customRole.getIncludedPermissions(),
+          projectId);
+      iamCow.projects().roles().create("projects/" + projectId, request).execute();
+    } catch (GoogleJsonResponseException googleEx) {
+      // Because this step may run multiple times, we need to handle duplicate role creation.
+      // The project was retrieved from RBS earlier in this flight, so we assume that any conflict
+      // of role names must be due to duplicate step execution.
+      if (googleEx.getStatusCode() != HttpStatus.CONFLICT.value()) {
+        throw new RetryException(googleEx);
```

Comment:
What's the difference in throwing this exception and returning `STEP_RESULT_FAILURE_RETRY`?

Reply:
As far as I know they behave the same way. I don't know if there's a subtle difference or it's just style